File: host-libs/lua/loaders/dotnet.lua
--- .NET loader registration for polyplug.
local ffi = require("ffi")

local ok = pcall(ffi.cdef, [[
    typedef struct { const uint8_t* ptr; size_t len; } PolyplugDotnetConfig;
    void* polyplug_dotnet_loader_create(const PolyplugDotnetConfig* cfg);
    void  polyplug_dotnet_loader_free(void* ptr);
    uint32_t polyplug_runtime_register_loader(void* rt, void* loader);
]])

local _lib = nil
local function get_lib()
    if not _lib then
        _lib = ffi.load("polyplug_dotnet")
    end
    return _lib
end

local M = {}

function M.register(rt, min_framework)
    min_framework = min_framework or "10.0"
    local lib = get_lib()
    local cfg = ffi.new("PolyplugDotnetConfig", {
        ffi.cast("const uint8_t*", min_framework),
        #min_framework
    })
    local loader = lib.polyplug_dotnet_loader_create(cfg)
    if loader == nil then
        error("polyplug: dotnet loader create failed")
    end
    local polyplug_lib = ffi.load("polyplug")
    local err = polyplug_lib.polyplug_runtime_register_loader(rt, loader)
    if err ~= 0 then
        error("polyplug: dotnet loader register failed: " .. tostring(err))
    end
end

return M

File: host-libs/js/loaders/dotnet.ts
let _lib: Deno.DynamicLibrary<typeof DOTNET_SYMBOLS> | null = null;

const DOTNET_SYMBOLS = {
    polyplug_dotnet_loader_create: {
        parameters: ["pointer"] as const,
        result: "pointer" as const,
    },
} satisfies Deno.ForeignLibraryInterface;

function getLib() {
    if (!_lib) {
        const libPath = Deno.env.get("POLYPLUG_DOTNET_LIB")
            ?? "libpolyplug_dotnet.so";
        _lib = Deno.dlopen(libPath, DOTNET_SYMBOLS);
    }
    return _lib;
}

export function registerDotnetLoader(
    rt: Deno.PointerValue,
    registerFn: (rt: Deno.PointerValue, loader: Deno.PointerValue) => number,
    minFramework: string = "10.0"
): void {
    const lib = getLib();
    const encoded = new TextEncoder().encode(minFramework);
    const cfgBuf = new Uint8Array(16); // ptr (8) + len (8)
    const ptr = Deno.UnsafePointer.of(encoded);
    const view = new DataView(cfgBuf.buffer);
    view.setBigUint64(0, BigInt(Deno.UnsafePointer.value(ptr)), true);
    view.setBigUint64(8, BigInt(encoded.length), true);
    const cfgPtr = Deno.UnsafePointer.of(cfgBuf);
    const loaderPtr = lib.symbols.polyplug_dotnet_loader_create(cfgPtr);
    if (loaderPtr === null) {
        throw new Error("polyplug: dotnet loader create failed");
    }
    const err = registerFn(rt, loaderPtr);
    if (err !== 0) {
        throw new Error(`polyplug: dotnet loader register failed: ${err}`);
    }
}
